### DISCUSSION TITLE
Is thread-local backend is safe with garbage collector?

### DISCUSSION THREAD

**Discussion by olii** [celery/celery#7014] [Q&A]
I am opening this issue as a question because I am not sure if it is a real issue or not.

I spent some time diagnosing memory leaks and I noticed a potential problem with the threaded pool executor and the thread-local storage.

When the `AsyncResult` object is collected by GC and `__del__()` is called then a backend is called. `backend` is a value that should be local to a thread but I am not sure in which thread GC is run. If the backend is just being used by some worker thread but the main thread runs GC then there may be a race condition.

https://github.com/celery/celery/blob/e2e3e95bf8ac9f85e1ee91753602c47bac878380/celery/result.py#L389-L392

[a comment by an automated account is not shown]

**Comment by auvipy** [comment]
hey, thanks for opening the issue and sharing your concern, did you try it yourself? if not can you give it a try ans share output?

**Comment by olii** [comment]
I didn't find the time to experiment with it.

**Comment by olii** [comment]
I got some time and conducted a few experiments. From my tests, it is apparent that the thread that runs `__del__()` can be any of the running threads. This can result in invalid thread-local variables being used in the subsequent code and can possibly be a source of deadlocks, connection interrupts, or memory leaks. More, any exception that is raised in the `__del__()` method cannot be caught by a standard exception handler, but one must register a hook [`sys.unraisablehook`](https://docs.python.org/3/library/sys.html#sys.unraisablehook).

Here is a link to docs: https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__
Here, they advise not to use `__del__()` for cleaning up resources: https://www.pythontutorial.net/python-oop/python-__del__/

 In our code, we use Redis and Celery canvas. Sometimes `AsyncResult` creates a pubsub subscription on a Redis socket. The subscription is freed in the `__del__()` method once the garbage collector decides so. We use `pymongo` which creates at least one background thread that runs in the process. This creates a window of opportunity for two threads to enter the same connection with different local storage. It may not be apparent in simple workers where only a single thread exists.

Celery heavily counts on `__del__()` and uses it for resource management. It is deep in Celery. I am afraid it can't be fixed easily. A major rewrite of Celery internals is necessary.

**Reply by auvipy** [reply]
thanks for your feedback. we will look further into it and see how we could improve celery internals
